native/dlahqr.go

```diff
@@ -0,0 +1,421 @@
+				h21 = s
+				h22 = h11
+			default: // Prepare to use Francis' double shift (i.e.,
+				// 2nd degree generalized Rayleigh quotient).
+				h11 = h[(i-1)*ldh+i-1]
+				h21 = h[(i-0)*ldh+i-1]
+				h12 = h[(i-1)*ldh+i]
+				h22 = h[(i-0)*ldh+i]
+			}
+			s := math.Abs(h11) + math.Abs(h12) + math.Abs(h21) + math.Abs(h22)
+			var (
+				rt1r, rt1i float64
+				rt2r, rt2i float64
+			)
+			if s != 0 {
+				h11 /= s
+				h21 /= s
+				h12 /= s
+				h22 /= s
+				tr := (h11 + h22) / 2
+				det := (h11-tr)*(h22-tr) - h12*h21
+				rtdisc := math.Sqrt(math.Abs(det))
+				if det >= 0 {
+					// Complex conjugate shifts.
+					rt1r = tr * s
+					rt2r = rt1r
+					rt1i = rtdisc * s
+					rt2i = -rt1i
+				} else {
+					// Real shifts (use only one of them).
+					rt1r = tr + rtdisc
+					rt2r = tr - rtdisc
+					if math.Abs(rt1r-h22) <= math.Abs(rt2r-h22) {
+						rt1r *= s
+						rt2r = rt1r
+					} else {
+						rt2r *= s
+						rt1r = rt2r
+					}
+					rt1i = 0
+					rt2i = 0
+				}
+			}
+
+			// Look for two consecutive small subdiagonal elements.
+			var m int
+			for m = i - 2; m >= l; m-- {
+				// Determine the effect of starting the
+				// double-shift QR iteration at row m, and see
+				// if this would make H[m,m-1] negligible. The
+				// following uses scaling to avoid overflows and
+				// most underflows.
+				h21s := h[(m+1)*ldh+m]
+				s := math.Abs(h[m*ldh+m]-rt2r) + math.Abs(rt2i) + math.Abs(h21s)
+				h21s /= s
+				v1 := h21s*h[m*ldh+m+1] + (h[m*ldh+m]-rt1r)*((h[m*ldh+m]-rt2r)/s) - rt2i/s*rt1i
+				v2 := h21s * (h[m*ldh+m] + h[(m+1)*ldh+m+1] - rt1r - rt2r)
+				v3 := h21s * h[(m+2)*ldh+m+1]
+				s = math.Abs(v1) + math.Abs(v2) + math.Abs(v3)
+				v1 /= s
+				v2 /= s
+				v3 /= s
+				if m == l {
+					break
+				}
+				dsum := math.Abs(h[(m-1)*ldh+m-1]) + math.Abs(h[m*ldh+m]) + math.Abs(h[(m+1)*ldh+m+1])
+				if math.Abs(h[m*ldh+m-1])*(math.Abs(v2)+math.Abs(v3)) <= ulp*math.Abs(v1)*dsum {
+					break
+				}
+			}
+
+			// Double-shift QR step.
+			var v [3]float64
+			for k := m; k < i; k++ {
+				// The first iteration of this loop determines a
+				// reflection G from the vector V and applies it
+				// from left and right to H, thus creating a
+				// non-zero bulge below the subdiagonal.
+				//
+				// Each subsequent iteration determines a
+				// reflection G to restore the Hessenberg form
+				// in the (k-1)th column, and thus chases the
+				// bulge one step toward the bottom of the
+				// active submatrix. nr is the order of G.
+
+				nr := min(3, i-k+1)
+				if k > m {
+					bi.Dcopy(nr, h[k*ldh+k-1:], ldh, v[:], 1)
+				}
+				var t0 float64
+				v[0], t0 = impl.Dlarfg(nr, v[0], v[1:], 1)
+				if k > m {
+					h[(k+0)*ldh+k-1] = v[0]
+					h[(k+1)*ldh+k-1] = 0
+					if k < i-1 {
+						h[(k+2)*ldh+k-1] = 0
+					}
+				} else if m > l {
+					// Use the following instead of H[k,k-1]
+					// = -H[k,k-1] to avoid a bug when v(2)
+					// and v(3) underflow.
+					h[k*ldh+k-1] *= 1 - t0
+				}
+				v1 := v[1]
```

Comment:
What is the reason for this? and the v2 assignment below? I don't see v being modified here.